source/common/router/router.cc

```diff
@@ -518,7 +550,27 @@ void Filter::onUpstreamHeaders(Http::HeaderMapPtr&& headers, bool end_stream) {
   upstream_request_->upstream_canary_ =
       (headers->EnvoyUpstreamCanary() && headers->EnvoyUpstreamCanary()->value() == "true") ||
       upstream_request_->upstream_host_->canary();
-  chargeUpstreamCode(*headers, upstream_request_->upstream_host_);
+  chargeUpstreamCode(*headers, upstream_request_->upstream_host_, false);
+  // We need to defer gRPC success until after we have processed
+  // grpc-status in the trailers.
+  const uint64_t response_code = Http::Utility::getResponseStatus(*headers);
+  if (!Http::CodeUtility::is5xx(response_code)) {
```

Comment:
nit: is it possible to reduce mega nesting here for readability possible by factoring out to a different function?